experimental/packages/opentelemetry-sdk-metrics-base/src/state/MetricStorageRegistry.ts

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { MetricStorage } from './MetricStorage';
+import { isDescriptorCompatibleWith } from '../InstrumentDescriptor';
+import * as api from '@opentelemetry/api';
+import { Maybe } from '../utils';
+import { getConflictResolutionRecipe, getIncompatibilityDetails } from '../view/RegistrationConflicts';
+
+/**
+ * Internal class for storing {@link MetricStorage}
+ */
+export class MetricStorageRegistry {
+  private readonly _metricStorageRegistry = new Map<string, MetricStorage[]>();
+
+  getStorages(): MetricStorage[] {
+    const storages = [];
+    for (const metricStorages of Array.from(this._metricStorageRegistry.values())) {
+      storages.push(...metricStorages);
+    }
+
+    return storages;
+  }
+
+  register<T extends MetricStorage>(storage: T): Maybe<T> {
+    const expectedDescriptor = storage.getInstrumentDescriptor();
+    const existingStorages = this._metricStorageRegistry.get(expectedDescriptor.name);
+
+    // Add storage if it does not exist.
+    if (existingStorages === undefined) {
+      this._metricStorageRegistry.set(expectedDescriptor.name, [storage]);
+      return storage;
+    }
+
+    const compatibleStorages = [];
+
+    for (const existingStorage of existingStorages) {
+      const existingDescriptor = existingStorage.getInstrumentDescriptor();
+
+      if (isDescriptorCompatibleWith(existingDescriptor, expectedDescriptor)) {
+        // Use the longer description if it does not match.
+        if (existingDescriptor.description !== expectedDescriptor.description) {
+          if (expectedDescriptor.description.length > existingDescriptor.description.length) {
+            existingStorage.updateDescription(expectedDescriptor.description);
+          }
+
+          api.diag.warn('A view or instrument with the name ',
+            expectedDescriptor.name,
+            ' has already been registered, but has a different description and is incompatible with another registered view.\n',
+            'Details:\n',
+            getIncompatibilityDetails(existingDescriptor, expectedDescriptor),
+            'The longer description will be used.\nTo resolve the conflict:',
+            getConflictResolutionRecipe(existingDescriptor, expectedDescriptor));
+        }
+        // Storage is fully compatible.
+        compatibleStorages.push(existingStorage as T);
```

Comment:
It seems like we can early break here so that we don't have to iterate over the remaining items. The code comment below also suggests that `compatibleStorages` contains at most 1 element.

Reply:
Yes, I was doing that initially. But I was trying to satisfy the following Spec statements:
> The implementation SHOULD warn about duplicate instrument registration conflicts after applying View configuration.

> If the potential conflict involves instruments that can be distinguished by a supported View selector (e.g., instrument type) a View recipe SHOULD be printed advising the user how to avoid the warning by renaming one of the conflicting instruments.

> Generally, for potential conflicts involving an identifying property (i.e., all properties except description), the producer SHOULD inform the user of a semantic error and pass through conflicting data.

Which seemed to me that it is necessary to continue iterating, even though it is impossible that there will be another compatible storage that already exists at this point. It could be that the first `MetricStorage` is compatible, but that there are others that are not. This guarantees that there will be warnings for (all) duplicates, and View recipes for every potential conflict. 

I'm not sure if I'm over-interpreting the spec, so if there's differing opinions I'd be happy to change it. :slightly_smiling_face:

Reply:
Sounds good to me. However, it seems like we don't need `compatibleStorages` to be an array, just a slot like `let compatibleStorage: MetricStorage | null` is good? This may suggest that we can have only one compatible storage for each instrumentation, and don't need to explain that "compatibleStorages may contain 1 element at most".

Reply:
Yes, that's a good idea. Fixed in [2488a4d](https://github.com/open-telemetry/opentelemetry-js/pull/2734/commits/2488a4d2ed441fe4dc6cc6248613c3eb8f1a9d41).